sdk/authentication.py

```diff
@@ -0,0 +1,123 @@
+        if not token_provider:
+            raise Exception("The authenticator needs a token provider.")
+        self.token_provider = token_provider
+
+    def __call__(self, r: PreparedRequest) -> PreparedRequest:
+        r.headers["Authorization"] = f"Bearer {self.token_provider.get_token()}"
+        return r
+
+
+class TokenProviderError(Exception):
+    pass
+
+
+class TokenProvider(ABC):
+    __token_cache: Optional[tuple[datetime, str]] = None
+    expiry_window: timedelta = timedelta(minutes=1.0)
+
+    @abstractmethod
+    def _query_token(self) -> dict[str, Any]:
+        raise NotImplementedError
+
+    def get_token(self) -> str:
+        if self.__token_cache:
+            expiry, token = self.__token_cache
+            if expiry - self.expiry_window > datetime.now(tz=timezone.utc):
+                return token
+
+        self.__token_cache = self._refresh_token_cache()
+        return self.__token_cache[1]
+
+    def _refresh_token_cache(self) -> tuple[datetime, str]:
+        try:
+            token_response = self._query_token()
+            expiry = self._extract_expiry(token_response)
+
+            if expiry is None:
+                expiry = datetime.now(tz=timezone.utc)
+
+            return (expiry, token_response["access_token"])
+        except HTTPError as err:
+            raise TokenProviderError(err)
+
+    @staticmethod
+    def _extract_expiry(token_response: dict[str, Any]) -> datetime | None:
+        """Extracts the expiry datetime from the token response.
+
+        Assumes that the actual token provided is correct, but might be in an
+        unexpected format
+
+        Args:
+            token_response: A valid token response dictionary.
+
+        Returns:
+            The time the token will expire, or None if it can't be calculated
+        """
+        expires_in = token_response.get("expires_in", None)
+        if expires_in:
+            return datetime.now(tz=timezone.utc) + timedelta(seconds=float(expires_in))
+
+        try:
+            # We assume the token is valid, but might not be in an expected format
+            token = token_response.get("access_token")
+
+            if (isinstance(token, str)and "." in token):
+                token_timestamp = decode(
+                        token,
+                        algorithms=["HS256"],
+                        options={"verify_signature": False},
+                    ).get("exp")
+                if token_timestamp:
+                    return datetime.fromtimestamp(token_timestamp, tz=timezone.utc)
+
+        except DecodeError:
+            # The token is not in a format that could be parsed as a JWT
+            pass
+        return None
+
+
+class Auth0TokenProvider(TokenProvider):
+    def __init__(self, username: str, password: str):
+        self.username = username
+        self.password = password
+
+        #   This is public, so we can put it in the code here.
+        self.public_client_id = "5QtfSu1UV118Iz6By6IJRSNoDrLbAiOv"
```

Comment:
This depends on the auth0 tenant and thus cannot be hardcoded if we want to make the auth0 endpoint configurable.

I'm not sure sure it would fit well inside the current `Endpoints` class, the naming doesn't sound very adequate.. Maybe it could be fine for now to add it there, i.e. make auth0 a dict  (or dataclass) with a domain key and a client_id key

Reply:
Does the `public_client_id` change ? It's my bad then, can you send me tomorrow the real `public_client_id` of the production ? I don't think I have access to it.

But yeah, if we don't want to hardcode it, I think, as I said in another comment, we should maybe make a system for the SDK to connect seemlessly to either `local | dev | preprod | prod` environment

Reply:
i changed the public client id to the prod one.
but indeed, we should in another PR make it configurable